### PR TITLE
(prometheus) adhoc filter support

### PR DESCRIPTION
**Pull request by mtanda**
Currently, Prometheus doesn't provide API to get all label keys.
I add label key input box to datasource setting page.

I'm not sure "expanding adhoc variable in replace function" is correct.
(the code is here, https://github.com/grafana/grafana/pull/6138/files#diff-ece81ff260bfcb9f65df22b3275da7aeR157)

If there is more appropriate way, I'll change the implementation.

https://github.com/grafana/grafana/issues/6038

### CODE REVIEW

File: public/app/features/templating/templateSrv.js [hunk 1]
```diff
@@ -154,6 +154,13 @@ function (angular, _, kbn) {
           }
         }
 
+        var adhocVar = _.find(self._adhocVariables, function(v) {
```
Comment (torkelo) [inline public/app/features/templating/templateSrv.js]: this should not be here, the filter should be added to the query automatically in the datasource query code (see elasticsearch or InfluxDB for example)

File: public/app/plugins/datasource/prometheus/partials/config.html [hunk 2]
```diff
@@ -1,3 +1,14 @@
 <datasource-http-settings current="ctrl.current" suggest-url="http://localhost:9090">
 </datasource-http-settings>
 
+<h3 class="page-heading">Prometheus details</h3>
+
+<div class="gf-form-group max-width-30">
+	<div class="gf-form">
+		<label class="gf-form-label width-13">Label Keys</label>
+		<input type="text" class="gf-form-input max-width-18" ng-model='ctrl.current.jsonData.labelKeys' placeholder="job,instance"></input>
```
Comment (torkelo) [inline public/app/plugins/datasource/prometheus/partials/config.html]: I do not think this global label keys option is a good idea. 
Think the proper solution is to
1) Add a metric/measurement input option the Ad Hoc Filter Variable options (that can limit the tag keys), will be optional for ES/InfluxDB but requrired for Prometheus. 

2) Wait until Prometheus implements a way to get all keys

**Review by torkelo — CHANGES_REQUESTED** [review]

File: public/app/features/templating/templateSrv.js (hunk 1 above)
Comment (mtanda) [inline public/app/features/templating/templateSrv.js]: Okay, I'll check it.

File: public/app/plugins/datasource/prometheus/partials/config.html (hunk 2 above)
Comment (mtanda) [inline public/app/plugins/datasource/prometheus/partials/config.html]: 1) is good choice for me. I'll try it.

File: public/app/features/templating/templateSrv.js (hunk 1 above)
Comment (mtanda) [inline public/app/features/templating/templateSrv.js]: Fixed this.

File: public/app/plugins/datasource/prometheus/partials/config.html (hunk 2 above)
Comment (mtanda) [inline public/app/plugins/datasource/prometheus/partials/config.html]: @torkelo I can't find the "metric/measurement input option" in template variable editor. Would you tell me where the option is exist?

Comment (torkelo) [inline public/app/plugins/datasource/prometheus/partials/config.html]: there is no such option yet, has be added

**Comment by bergquist** [comment]
ref https://github.com/prometheus/prometheus/issues/2038

File: public/app/plugins/datasource/prometheus/partials/config.html (hunk 2 above)
Comment (mtanda) [inline public/app/plugins/datasource/prometheus/partials/config.html]: I change to gather label keys from query response.
User don't need to set label key manually.

Comment (torkelo) [inline public/app/plugins/datasource/prometheus/partials/config.html]: how is that possible? dont you need a metric name for a query?

**Comment by torkelo** [comment]
>Please see this query.
It is normal case, all metrics in query doesn't have some of labels.
The method_code:http_errors:rate5m has code labels, but method:http_requests:rate5m is not.

This feature is not built for that case then. For that case you can use normal query variable. The Ad hoc filters is for exploratory dashboards where you can filter all data by ad hoc key/value pair.s

**Comment by mtanda** [comment]
hmm, it is quiet useful if we can build filter part in adhoc filter UI.
But, I understand why you don't want to embed template variable in query expression.
User should start to make adhoc filter without opening query editor.

The query doesn't have "on" or "ignoring" clause, it is safe to add filter part to all metrics.
And, if there are "on" or "ignoring" clause, the adhoc filter shouldn't show label key in "ignoring" clause or something.

The query parser provide such information, it is possible to adhoc filter support in Prometheus.

File: public/app/plugins/datasource/prometheus/partials/config.html (hunk 2 above)
Comment (mtanda) [inline public/app/plugins/datasource/prometheus/partials/config.html]: If it is OK to show all label keys in UI, no need to metric name.

If user start to edit adhoc filter, the dashboard might have some graph panel, and datasource plugin already receive query response.
Gathering label keys in the response, and provide label keys to adhoc filter UI is good choice, I think.
The label keys in query response is mostly usable in the dashboard.

Comment (mtanda) [inline public/app/plugins/datasource/prometheus/partials/config.html]: I reconsider about the possibility to get all label keys from query response.
And it is NO. Some of the label keys can get from response.
But, if the query use "count" or aggregate function, the labels which metrics has, is not appeared in query response...

Gathering label keys from response could improve coverage of label keys.
But, need to input box too. I'll add it.

**Comment by mtanda** [comment]
I misunderstand about this part.
```
All job="node" is filter part.
So it is meaningless to find first "{".
```

To use adhoc filter, don't need to add adhoc filter for all PromQL filter part.
Add first one is enough, other metrics automatically follow first one's filter.
Some of query, the metrics order is important (like "group_left" or "group_right"), but it can be rewritable, it's OK.

**Comment by mtanda** [comment]
https://prometheus.io/docs/querying/operators/#arithmetic-binary-operators
> Between two instant vectors, a binary arithmetic operator is applied to each entry in the left-hand-side vector and its matching element in the right hand vector. The result is propagated into the result vector and the metric name is dropped. Entries for which no matching entry in the right-hand vector can be found are not part of the result.

File: public/app/plugins/datasource/prometheus/partials/config.html (hunk 2 above)
Comment (mtanda) [inline public/app/plugins/datasource/prometheus/partials/config.html]: Added "metric" input box.

**Comment by mtanda** [comment]
This PR is ready to re-review.

**Comment by filippog** [comment]
Hi, has there been any progress on this feature?
It seems a really useful feature to be able to ad-hoc explore prometheus dashboards too

**Comment by mtanda** [comment]
I wait for Prometheus query editor refinement.
https://github.com/grafana/grafana/pull/6138#commitcomment-19208398

When it is ready, I'll fix conflict.

**Comment by filippog** [comment]
> I wait for Prometheus query editor refinement.
> #6138 (comment)
> 
> When it is ready, I'll fix conflict.

Thanks @mtanda ! I can't find a specific GH issue for that, is there one open already I could follow?

**Comment by mtanda** [comment]
@filippog I didn't create GH issue.
If you create the issue with describing your use case, it is greatly helpful :-)

**Comment by filippog** [comment]
> If you create the issue with describing your use case, it is greatly helpful :-)

re-reading this issue it looks like my use case is exactly like yours, thanks!

**Comment by torkelo** [comment]
closing this for now as Prometheus api does not currently support the required api, and it could be better implemented after we have new query editor & model

**Pull request closed without merging**